src/Planner/PlannerJoins.cpp

```diff
@@ -385,6 +600,203 @@ void buildJoinClause(
+                    result.insert(result.end(), std::make_move_iterator(child_res.begin()), std::make_move_iterator(child_res.end()));
+                }
+                it++;
+
+                for (; it != arguments.end(); it++)
+                {
+                    auto & child_res = built_clauses.at(it->get());
+                    result = makeCrossProduct(result, child_res);
+                }
+            }
+
+            built_clauses.emplace(node.get(), std::move(result));
+        }
+        else
+        {
+            nodes_to_process.pop();
+            JoinClauses clauses;
+            clauses.emplace_back();
+            buildSimpleJoinClause(
+                left_dag,
+                right_dag,
+                joined_dag,
+                planner_context,
+                node,
+                left_table_expressions,
+                right_table_expressions,
+                join_node,
+                clauses.front());
+
+            built_clauses.emplace(node.get(), std::move(clauses));
+        }
+    }
+    return std::move(built_clauses.at(join_expression.get()));
+}
+
+std::pair<JoinClauses, bool /*is_inequal_join*/> buildAllJoinClauses(
+    ActionsDAG & left_join_actions,
+    ActionsDAG & right_join_actions,
+    ActionsDAG & post_join_actions,
+    const PlannerContextPtr & planner_context,
+    const QueryTreeNodePtr & join_expression,
+    const TableExpressionSet & join_left_table_expressions,
+    const TableExpressionSet & join_right_table_expressions,
+    const JoinNode & join_node,
+    const FunctionNode & function_node)
+{
+    const auto & join_algorithms = planner_context->getQueryContext()->getSettingsRef()[Setting::join_algorithm];
+    const auto is_hash_join_enabled = TableJoin::isEnabledAlgorithm(join_algorithms, JoinAlgorithm::HASH)
+        || TableJoin::isEnabledAlgorithm(join_algorithms, JoinAlgorithm::AUTO);
+    if (is_hash_join_enabled && planner_context->getQueryContext()->getSettingsRef()[Setting::allow_general_join_planning])
+    {
+        auto join_clauses = buildJoinClauses(
+            left_join_actions,
+            right_join_actions,
+            post_join_actions,
+            planner_context,
+            join_expression,
+            join_left_table_expressions,
+            join_right_table_expressions,
+            join_node);
+
+        const auto has_residual_filters = std::any_of(
+            join_clauses.begin(),
+            join_clauses.end(),
+            [](const JoinClause & clause) { return !clause.getResidualFilterConditionNodes().empty(); });
+
+        return std::make_pair(std::move(join_clauses), has_residual_filters);
+    }
+
+    bool has_residual_filters = false;
+    JoinClauses join_clauses;
+    const auto & function_name = function_node.getFunction()->getName();
+    if (function_name == "or")
+    {
+        for (const auto & child : function_node.getArguments())
+        {
+            join_clauses.emplace_back();
+
+            buildJoinClause(
```

Comment:
I think we don't do recursive call here and `buildJoinClause` only calls itself in case of `AND` functions. If you meant that, then I think we cannot easily replace that, as flattening the conditions only happens in `LogicalExpressionOptimizerPass` if and only when we can extract some subexpressions. But e.g.: `A & (B & C)` won't be flattened, because there is no common expression.